### ISSUE TITLE
More consistent signatures for filters and pseudos?

### ISSUE THREAD

**Issue by chocolateboy** [fb55/css-select#104]
There's a difference between the way the adapter is passed to pseudos:

```javascript
pseudo(elem, adapter, subselect)
```
and the way it's passed to filters:

```javascript
filter(next, subselect, { adapter, ... }, context)
```

This is both confusing (I can't think of any other APIs which present the same parameter in two different ways like this), and limiting, as it means `compile` options can't be passed down to pseudos.

I realize it's too late for css-select v2.0, but, given that it's not too late for Cheerio v1.0 ([yet](https://github.com/cheeriojs/cheerio/pull/1145)), how about making the API more consistent and flexible across these related function types?

At least by 1) using the options object for both:

```javascript
filter(next, subselect, options, context)
pseudo(elem, options, subselect)
```

And possibly by 2) making the parameter ordering more consistent as well e.g.:

```javascript
filter(next, context, subselect, options)
pseudo(elem, subselect, options)
```

or:

```javascript
filter(subselect, options, next, context)
pseudo(subselect, options, elem)
```

Typically, this kind of shared parameter would be defined on `this` rather than passed around everywhere. This could be done by 3) making CSSselect a constructor and moving the function it currently exports to e.g. a `select` method:

```javascript
const cssSelect = new CSSselect({ adapter })
cssSelect.select(...)
```

But even without going down that route (which would presumably entail a major rewrite), the adapter could still be 4) accessed via `this.adapter` e.g.:

```javascript
filter.call(options, subselect, next, context)
pseudo.call(options, subselect, elem)
```

**Comment by fb55** [comment]
PRs welcome! :)

**Comment by fb55** [comment]
Just saw that I closed #105 without a comment — super sorry about that.

As a late reply: I've just completed a TypeScript migration of this library, which should help with writing custom filters & pseudos. The current signature is pretty good at minimizing the number of skipped arguments, and I'd like to keep it in its current state.
